pkg/machineproviders/providers/openshift/machine/v1beta1/provider.go

```diff
@@ -128,16 +156,289 @@ type openshiftMachineProvider struct {
+
+	machineList := &machinev1beta1.MachineList{}
+	if err := m.client.List(ctx, machineList, &client.ListOptions{LabelSelector: selector}); err != nil {
+		return nil, fmt.Errorf("failed to list machines: %w", err)
+	}
+
+	correctFormat := isMachinesNamesFormatCorrect(machineList.Items)
+
+	failureDomains, err := providerconfig.ExtractFailureDomainsFromMachines(machineList.Items)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract failure domains from machines: %w", err)
+	}
+
+	for i, machine := range machineList.Items {
+		machineInfo, err := m.generateMachineInfo(machine, failureDomains[i], correctFormat, logger)
+		if err != nil {
+			return nil, err
+		}
+
+		machineInfos = append(machineInfos, *machineInfo)
+
+		nodeName := ""
+		if machineInfo.NodeRef != nil {
+			nodeName = machineInfo.NodeRef.ObjectMeta.Name
+		}
+
+		logger.V(4).Info(
+			"Gathered Machine Info",
+			"machineName", machine.GetName(),
+			"nodeName", nodeName,
+			"index", int(machineInfo.Index),
+			"ready", machineInfo.Ready,
+			"needsUpdate", machineInfo.NeedsUpdate,
+			"errorMessage", machineInfo.ErrorMessage,
+		)
+	}
+
+	return machineInfos, nil
+}
+
+// generateMachineInfo creates a MachineInfo object for a given machine.
+func (m *openshiftMachineProvider) generateMachineInfo(machine machinev1beta1.Machine, failureDomain failuredomain.FailureDomain, correctFormat bool, logger logr.Logger) (*machineproviders.MachineInfo, error) {
+	machineRef := getMachineRef(machine)
+
+	nodeRef := getNodeRef(machine)
+
+	errorMessage := getMachineErrorMessage(machine)
+
+	index, indexFound := m.failureDomainToIndex(failureDomain)
+	if !indexFound && !correctFormat {
+		// When the machine names do not fit the pattern, and the failure domains are
+		// not recognised, returns an error.
+		logger.Error(errCouldNotDetermineMachineIndex,
+			"Could not gather Machine Info",
+		)
+
+		return nil, errCouldNotDetermineMachineIndex
+	}
+
+	// If machine name index and domain index are different, we should use the machine
+	// name index in the result, but only if the name is correctly formatted
+	if correctFormat {
+		// We can ignore the error here as we use machineNameIndex only when we are sure that all
+		// machines have the correct name format
+		machineNameIndex, _ := strconv.Atoi(machine.Name[strings.LastIndex(machine.Name, "-")+1:]) //nolint:gosec
+
+		if int32(machineNameIndex) != index {
+			index = int32(machineNameIndex)
+		}
+	}
+
+	providerConfig, err := providerconfig.NewProviderConfigFromMachine(machine)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create provider config from machine: %w", err)
+	}
+
+	templateProviderConfig, err := m.providerConfig.InjectFailureDomain(failureDomain)
+	if err != nil {
+		return nil, fmt.Errorf("error injecting failure domain into provider config: %w", err)
+	}
+
+	configsEqual, err := templateProviderConfig.Equal(providerConfig)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compare provider configs: %w", err)
```

Comment:
this error seems a little cryptic, is there something that the user could do here to fix things?

Reply:
Maybe this will help a little, not sure what exactly the errors coming out may be though
```suggestion
		return nil, fmt.Errorf("could not compare existing and desired provider configs: %w", err)
```